--- package/feeds/packages/runc/Makefile
+++ package/feeds/packages/runc/Makefile
@@ -80,6 +80,7 @@
 # Avoid installing binaries
 define Build/InstallDev
 	$(call Build/Compile/Default,clean)
+	rm -f $(STAMP_BUILT)
 	$(call GoPackage/Build/InstallDev,$(1))
 endef

--- package/feeds/packages/containerd/Makefile
+++ package/feeds/packages/containerd/Makefile
@@ -75,6 +75,7 @@
 # Avoid installing binaries
 define Build/InstallDev
 	$(call Build/Compile/Default,clean)
+	rm -f $(STAMP_BUILT)
 	$(call GoPackage/Build/InstallDev,$(1))
 endef
