--- package/openwrt-packages/OpenClash/luci-app-openclash/files/etc/init.d/openclash
+++ package/openwrt-packages/OpenClash/luci-app-openclash/files/etc/init.d/openclash
@@ -620,5 +620,11 @@
 fi
 }
 
+redir_bypass()
+{
+  local ipcidr="$1"
+  iptables -t nat -A openclash -s $ipcidr -j RETURN
+}
+
 start()
 {
@@ -764,6 +770,8 @@
              
              if [ -z "$en_mode_tun" ]; then
                 iptables -t nat -N openclash
+                config_load "openclash"
+                config_list_foreach 'config' 'redir_bypass' redir_bypass
                 iptables -t nat -A openclash -d 0.0.0.0/8 -j RETURN
                 iptables -t nat -A openclash -d 10.0.0.0/8 -j RETURN
                 iptables -t nat -A openclash -d 127.0.0.0/8 -j RETURN
