--- package/feeds/packages/netdata/Makefile
+++ package/feeds/packages/netdata/Makefile 
@@ -31,7 +31,7 @@ include $(INCLUDE_DIR)/package.mk
 define Package/netdata
   SECTION:=admin
   CATEGORY:=Administration
-  DEPENDS:=+zlib +libuuid +libuv +libmnl +libjson-c
+  DEPENDS:=+zlib +libuuid +libuv +libmnl +libjson-c +libopenssl +liblz4 +libJudy
   TITLE:=Real-time performance monitoring tool
   URL:=https://www.netdata.cloud/
 endef
@@ -54,8 +54,6 @@ 	--disable-x86-sse \
 	--enable-lto \
 	--disable-ebpf \
 	--without-libcap \
-	--disable-https \
-	--disable-dbengine \
 	--disable-compression \
 	--disable-plugin-nfacct \
 	--disable-plugin-freeipmi \
